### DISCUSSION TITLE
How to clear history and merge commits

### DISCUSSION THREAD

**Discussion by schnek** [AscEmu/AscEmu#1039] [Q&A] [answered]
First, let's find out how many commits need to be merged.&#128016;
This command will show you what commits you have added compared to the develop branch:

`git cherry -v develop`

![Image](https://user-images.githubusercontent.com/1216225/173328433-9d002922-83c5-4ea0-9cbc-b789b54bda07.png)

how many changes

`git cherry -v develop | wc -l`

![Image](https://user-images.githubusercontent.com/1216225/173328803-97ccaaa2-d229-4277-9a77-64f56d5858e3.png)

I have 2 commits. Now I want to rewrite history from HEAD~2, that is, from what was 2 commits ago. To do this, I do this:

`git rebase -i HEAD~2`

a text editor opens automatically.

![Image](https://user-images.githubusercontent.com/1216225/95241214-20f28700-0816-11eb-8c5e-a29e562675e5.png)

Here's a list of my commits and a big comment on what I can do. From the list of commands, you can see that we can use _squash_ or _fixup_ in order to glue commits. The first comes in handy when you want to change the commit message, and the second when you want to use the commit message of the first. So, to glue all the commits, I do like this:

`git push --force`

![Image](https://user-images.githubusercontent.com/1216225/173329897-aeeb93f3-68e5-41d9-b2fa-d021934d4c34.png)

before

![Image](https://user-images.githubusercontent.com/1216225/173330259-8b3c376e-9d55-4def-a9ff-f60be856c4e6.png)

after

![Image](https://user-images.githubusercontent.com/1216225/95242161-77ac9080-0817-11eb-92e0-ed9723262437.png)

remember delete cannot be restored

**Comment by Appled** [comment, marked as answer]
Or you can just do "git rebase -i develop" and it will show all commits that are not on develop branch 😃

Also, to note that after rebase and force push other users cannot pull your new changes if they had cloned your branch. They have to delete their local branch and fetch it again.

**Reply by schnek** [reply]
if you do not change the commits, then this is also possible^^
